File: src/env.ts
import type { ProxyConfig } from './types.ts';

function parsePort(raw: string | undefined): number {
  const port = Number(raw ?? 8787);
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new Error('PORT must be an integer between 1 and 65535');
  }
  return port;
}

function parseBoolEnv(name: string, raw: string | undefined): boolean {
  if (raw === undefined || !raw.trim()) return false;
  switch (raw.trim().toLowerCase()) {
    case '1':
    case 'true':
    case 'yes':
    case 'on':
      return true;
    case '0':
    case 'false':
    case 'no':
    case 'off':
      return false;
    default:
      throw new Error(`${name} must be a boolean value`);
  }
}

function parseNonNegativeIntegerEnv(name: string, raw: string | undefined): number {
  if (raw === undefined || !raw.trim()) return 0;
  const value = Number(raw.trim());
  if (!Number.isInteger(value) || value < 0) {
    throw new Error(`${name} must be a non-negative integer`);
  }
  return value;
}

function parseIntervalMsEnv(name: string, raw: string | undefined, fallback: number): number {
  if (raw === undefined || !raw.trim()) return fallback;
  const value = Number(raw.trim());
  if (!Number.isInteger(value) || value < 0) {
    throw new Error(`${name} must be a non-negative integer`);
  }
  return value;
}

function parseOptionalPositiveIntegerEnv(name: string, raw: string | undefined): number | null {
  if (raw === undefined || !raw.trim()) return null;
  const value = Number(raw.trim());
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`${name} must be a positive integer`);
  }
  return value;
}

function parseThresholdPercentEnv(name: string, raw: string | undefined): number {
  if (raw === undefined || !raw.trim()) return 90;
  const value = Number(raw.trim());
  if (!Number.isFinite(value) || value <= 0 || value >= 100) {
    throw new Error(`${name} must be a number greater than 0 and less than 100`);
  }
  return value;
}

function isTextFilePath(raw: string): boolean {
  return raw.trim().toLowerCase().endsWith('.txt');
}

function normalizeApiKeyEntry(raw: string): string {
  return raw.trim().replace(/,+$/, '').trim();
}

function parseApiKeys(raw: string | undefined): string[] {
  const source = raw?.trim() ?? '';
  const keys = isTextFilePath(source)
    ? Deno.readTextFileSync(source)
      .split(/\r?\n/)
      .map(normalizeApiKeyEntry)
      .filter(Boolean)
    : source.split(',').map(normalizeApiKeyEntry).filter(Boolean);
  if (keys.length === 0) throw new Error('OPENAI_API_KEY is required');
  return keys;
}

export function loadDotenvIntoEnv(path: string): void {
  const text = Deno.readTextFileSync(path);
  for (const rawLine of text.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith('#')) continue;
    const eq = line.indexOf('=');
    if (eq <= 0) continue;
    const key = line.slice(0, eq).trim();
    if (!key || key.toUpperCase().startsWith('CODEX_')) continue;
    let value = line.slice(eq + 1).trim();
    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      value = value.slice(1, -1);
    }
    Deno.env.set(key, value);
  }
}

export function applyLogArgsToEnv(args: string[]): void {
  let logDir: string | null = null;
  let sawLogArg = false;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--logs') {
      sawLogArg = true;
      const next = args[i + 1];
      if (next && !next.startsWith('--')) {
        logDir = next;
        i++;
      } else {
        logDir = 'logs';
      }
      continue;
    }
    if (arg === '--log-dir') {
      sawLogArg = true;
      const next = args[i + 1];
      if (!next || next.startsWith('--')) {
        throw new Error('missing value for --log-dir');
      }
      logDir = next;
      i++;
    }
  }

  if (!sawLogArg) {
    Deno.env.delete('HUBPROXY_LOG_DIR');
    return;
  }

  const normalized = logDir?.trim() || 'logs';
  Deno.env.set('HUBPROXY_LOG_DIR', normalized);
}

export function loadConfig(): ProxyConfig {
  const responsesBaseUrl = Deno.env.get('RESPONSES_BASE_URL');
  const chatBaseUrl = Deno.env.get('CHAT_BASE_URL');
  const defaultModel = Deno.env.get('DEFAULT_MODEL');
  const apiKeys = parseApiKeys(Deno.env.get('OPENAI_API_KEY'));
  const dataDir = Deno.env.get('DATA_DIR');
  const authToken = Deno.env.get('AUTH') ?? null;
  const accountEmail = Deno.env.get('ACCOUNT_EMAIL') ?? null;
  const accountName = Deno.env.get('ACCOUNT_NAME') ?? null;
  const accountPlanType = Deno.env.get('ACCOUNT_PLAN_TYPE') ?? null;
  const forceResponsesEnv = Deno.env.get('HUBPROXY_FORCE_RESPONSES') ??
    Deno.env.get('HUBPROXY_FORCE_RESPONESE');
  if (!chatBaseUrl) throw new Error('CHAT_BASE_URL is required');
  if (!defaultModel) throw new Error('DEFAULT_MODEL is required');
  if (!dataDir) throw new Error('DATA_DIR is required');
  return {
    port: parsePort(Deno.env.get('PORT')),
    host: Deno.env.get('HOST') ?? '0.0.0.0',
    authToken,
    accountEmail,
    accountName,
    accountPlanType,
    responsesBaseUrl: responsesBaseUrl ?? null,
    chatBaseUrl,
    forceChatCompletions: parseBoolEnv(
      'HUBPROXY_FORCE_CHAT_COMPLETIONS',
      Deno.env.get('HUBPROXY_FORCE_CHAT_COMPLETIONS'),
    ),
    forceResponses: parseBoolEnv('HUBPROXY_FORCE_RESPONSES', forceResponsesEnv),
    disablePromptInjection: parseBoolEnv(
      'HUBPROXY_DISABLE_PROMPT_INJECTION',
      Deno.env.get('HUBPROXY_DISABLE_PROMPT_INJECTION'),
    ),
    nvidiaCompat: parseBoolEnv('HUBPROXY_NVIDIA_COMPAT', Deno.env.get('HUBPROXY_NVIDIA_COMPAT')),
    isCloudflare: parseBoolEnv('IS_CF', Deno.env.get('IS_CF')),
    defaultModel,
    defaultApiKey: apiKeys[0],
    apiKeys,
    requestIntervalMs: parseNonNegativeIntegerEnv(
      'HUBPROXY_REQUEST_INTERVAL_MS',
      Deno.env.get('HUBPROXY_REQUEST_INTERVAL_MS'),
    ),
    needRetry: parseBoolEnv('NEED_RETRY', Deno.env.get('NEED_RETRY')),
    glmTryGetKey: parseBoolEnv('GLM_TRY_GET_KEY', Deno.env.get('GLM_TRY_GET_KEY')),
    glmKeyRefreshIntervalMs: parseIntervalMsEnv(
      'GLM_KEY_REFRESH_INTERVAL_MS',
      Deno.env.get('GLM_KEY_REFRESH_INTERVAL_MS'),
      600000,
    ),
    glmKeyFetchRetryCount: parseIntervalMsEnv(
      'GLM_KEY_FETCH_RETRY_COUNT',
      Deno.env.get('GLM_KEY_FETCH_RETRY_COUNT'),
      100,
    ),
    glmKeyFetchRetryDelayMs: parseIntervalMsEnv(
      'GLM_KEY_FETCH_RETRY_DELAY_MS',
      Deno.env.get('GLM_KEY_FETCH_RETRY_DELAY_MS'),
      30000,
    ),
    dataDir,
    customContextWindowTokens: parseOptionalPositiveIntegerEnv(
      'HUBPROXY_CONTEXT_WINDOW_TOKENS',
      Deno.env.get('HUBPROXY_CONTEXT_WINDOW_TOKENS'),
    ),
    contextCompactThresholdPercent: parseThresholdPercentEnv(
      'HUBPROXY_CONTEXT_COMPACT_THRESHOLD_PERCENT',
      Deno.env.get('HUBPROXY_CONTEXT_COMPACT_THRESHOLD_PERCENT'),
    ),
  };
}

File: .code_index/skeleton/src/env.py
from __future__ import annotations

from .types import ProxyConfig

# @origin src/env.ts:3
def parsePort(raw: str | None) -> float:
    Number(...)
    return Number.isInteger(...)
    raise Error(...)

# @origin src/env.ts:11
def parseBoolEnv(name: str, raw: str | None) -> bool:
    raw.trim(...)
    return toLowerCase(...)
    raise Error(...)

# @origin src/env.ts:29
def parseNonNegativeIntegerEnv(name: str, raw: str | None) -> float:
    raw.trim(...)
    Number(...)
    return Number.isInteger(...)
    raise Error(...)

# @origin src/env.ts:38
def parseIntervalMsEnv(name: str, raw: str | None, fallback: float) -> float:
    raw.trim(...)
    Number(...)
    return Number.isInteger(...)
    raise Error(...)

# @origin src/env.ts:47
def parseOptionalPositiveIntegerEnv(name: str, raw: str | None) -> float | None:
    raw.trim(...)
    Number(...)
    return Number.isInteger(...)
    raise Error(...)

# @origin src/env.ts:56
def parseThresholdPercentEnv(name: str, raw: str | None) -> float:
    raw.trim(...)
    Number(...)
    return Number.isFinite(...)
    raise Error(...)

# @origin src/env.ts:65
def parseModelMapEnv(name: str, raw: str | None) -> dict[str, str]:
    raw.trim(...)
    raw.split(...)
    rawEntry.trim(...)
    entry.indexOf(...)
    entry.slice(...)
    return trim(...)
    raise Error(...)

# @origin src/env.ts:85
def isTextFilePath(raw: str) -> bool:
    raw.trim(...)
    toLowerCase(...)
    return endsWith(...)

# @origin src/env.ts:89
def normalizeApiKeyEntry(raw: str) -> str:
    raw.trim(...)
    replace(...)
    return trim(...)

# @origin src/env.ts:93
def parseApiKeys(raw: str | None) -> list[str]:
    trim(...)
    isTextFilePath(...)
    Deno.readTextFileSync(...)
    split(...)
    map(...)
    filter(...)
    return source.split(...)
    raise Error(...)

# @origin src/env.ts:105
def loadDotenvIntoEnv(path: str) -> None:
    Deno.readTextFileSync(...)
    text.split(...)
    rawLine.trim(...)
    line.startsWith(...)
    line.indexOf(...)
    line.slice(...)
    trim(...)
    key.toUpperCase(...)
    startsWith(...)
    value.startsWith(...)
    value.endsWith(...)
    value.slice(...)
    Deno.env.set(...)

# @origin src/env.ts:125
def applyLogArgsToEnv(args: list[str]) -> None:
    next.startsWith(...)
    Deno.env.delete(...)
    trim(...)
    Deno.env.set(...)
    raise Error(...)

# @origin src/env.ts:162
def loadConfig() -> ProxyConfig:
    Deno.env.get(...)
    parseApiKeys(...)
    parsePort(...)
    parseBoolEnv(...)
    parseModelMapEnv(...)
    parseNonNegativeIntegerEnv(...)
    parseIntervalMsEnv(...)
    parseOptionalPositiveIntegerEnv(...)
    return parseThresholdPercentEnv(...)
    raise Error(...)
